File: 프로그래머스/2/84512. 모음 사전/모음 사전.py

# from itertools import product
# #product는 데카르트의 수를 구현한 메서드
# #product라는 python 메서드를 이용해서 중복순열 구현
# def solution(word):
#     permutation=[]
#     for j in range(1,6): #j는 몇 개의 글자를 뽑을것인가? -> 길이1~길이5인 단어들을 뽑을 것이기에 (1,6)반복 
#         for i in product(['A', 'E', 'I', 'O', 'U'], repeat=j):#길이1 중복순열, 길이2 중복순열 ... 
#             permutation.append(''.join(i))
#     # print(permutation)
#     answer=0
#     permutation.sort()
#     answer=permutation.index(word)+1
#     return answer


# DFS로 풀어보기

def solution(word):
    alphabets=['A', 'E', 'I', 'O', 'U']
    dict=[]
#     트리 모양으로도 그릴 수 있겠다 -> dfs로 순회하면서 dict에 있는지 확인하고 추가하는 과정 거치자!
#   dfs(value, count)
    def dfs(value, count):
        if count>5:
            return
        if value not in dict:
            dict.append(value)
        for j in alphabets:
            dfs(value+j,count+1)
    
    for alpha in alphabets:
        dfs(alpha,1)
    dict.sort()
    # print(dict)
    answer=0
    answer=dict.index(word)+1
    return answer

File: 프로그래머스/2/84512. 모음 사전/모음 사전.js
/*
단어 input 받아서 단어 철자로 만들 수 있는 모든 단어를 담은 사전에서 몇 번째 단어인지 return 

! 방법:
1. 단어를 array 형태로 나누어서 
2. 순열 구현해서 모든 경우의 수 모은다.
3. 경우의 수 사전에 맞게 오름차순 정렬 후 순서 구함

*/

function solution(word) {
    let answer = 0;
    let arr= ['A', 'E', 'I', 'O', 'U']
    let dictionary=[]
    for(let i=1; i<=5;i++){
        let permutationList=permutation(arr,i)
        // console.log(permutationList)
        dictionary=dictionary.concat(permutationList)
    }
    dictionary.sort()
    // console.log(dictionary)
    answer=dictionary.indexOf(word)+1
    return answer;
}

function permutation(array,length){
    // let visited= Array(array.length).fill(false)
    let result=[]
    function dfs(nowPerArr){
        if (nowPerArr.length==length){    //목표길이 달성시 빠져나오기
            result.push(nowPerArr.join(''))
            // console.log(nowPerArr)
            return
        }
        for(let j=0 ;j< 5;j++){  //
            // if(!visited[j]){
            //     visited[j]=true
            nowPerArr.push(array[j])
            dfs(nowPerArr)
            nowPerArr.pop()
            //     visited[j]=false
            // }
        }
    }
    dfs([])
    return result
}